### PR TITLE
[Feature] Route a single datasource across multiple databases

### PR DESCRIPTION
**Pull request by liuyufei9527**
## Why

The built-in benchmarks already intend "one datasource holds many databases" (BIRD's db_id is a database under a fixed datasource), but the implementation contradicted it: the glob parser flattened every matched file into its own top-level datasource, and the connection layer routed by datasource only. As a result a sqlite/duckdb glob could not select among its files, and the interactive `/database` switch silently kept querying the first file.

## Solution

Model a datasource as a connection profile that can serve multiple databases, routed by `(datasource, database)`:

- `DBManager.get_conn(datasource, database)` resolves and caches per `(datasource, database)`; a glob datasource's matched files are its databases, a server DB's `database` is cloned into the connection URI.
- DBFuncTool, nodes, benchmark, and the CLI all route through `get_conn`; the chat node rebinds its DB tools to the active database on `/database`.
- `schema_metadata` is initialised and scoped per `(datasource, database)`; semantic/metric stays datasource-level for all datasource types.
- `AgentConfig.list_databases(datasource)` enumerates a datasource's databases (glob files or the configured/default db).

Root-cause fix for the interactive sqlite `/database` switch: `BaseSqlConnector.database_name` is ContextVar-backed (per-thread / per-async-task). `SQLiteConnector`/`DuckdbConnector` set it via the property setter in `__init__`, which only populates the building context's ContextVar and never the context-independent `_default_database` fallback. Connectors are built once and cached, frequently in a worker thread (startup schema init, the CLI's `_init_connection` ThreadPoolExecutor), then read from the main thread — so `database_name` read back empty, leaving `current_db_name` empty after a switch. That broke the status bar, the `/databases` highlight, and chat tool routing (LLM `list_tables` hit the first glob file) while direct SQL and server DBs (built in the current context) worked. Fix: populate `_default_database` in `__init__` so the name is visible from every context; per-operation `switch_context` overrides still work.

## Test Cases

- Cross-thread regression tests for both connectors (`test_connector_sqlite.py`, `test_duckdb_connector.py`) asserting `database_name` survives a worker-thread build / main-thread read, including an explicit `database_name` override and the per-operation setter.
- Updated unit tests across `db_manager`, `database` func-tool, `metadata_commands`, `status_bar`, `agent_config`, `local_init`, and node tests for the `(datasource, database)` routing and one-datasource glob model.

## Backport

- [ ] release/0.3.1
- [ ] release/0.3.2
- [ ] release/0.3.3
## Release Notes

* **New Features**
  * Enhanced multi-database support enabling explicit selection and switching between databases within a single data source
  * Improved database context handling in CLI operations for more reliable datasource/database routing

* **Improvements**
  * Refined database selection logic for file-based datasources with multiple database files
  * Better database connection management and per-operation database routing

## Summary by CodeRabbit

* **New Features**
  * Multi-database routing: queries and per-task operations target a specific physical database within a shared datasource
  * Tools and agent nodes remember and bind to the selected active database for subsequent actions

* **Improvements**
  * Simplified datasource configuration and runtime database listing for file-pattern datasources
  * CLI, status bar, and metadata commands more consistently reflect the current database selection

* **Bug Fixes**
  * Per-task benchmark SQL execution now correctly targets each task’s configured database

### CODE REVIEW

File: datus/cli/service_manager.py [hunk 1]
```diff
@@ -286,12 +285,10 @@ def _save_configuration(self) -> bool:
...
                         "uri": db_config.uri,
                     }
-                    if db_config.logic_name and db_config.logic_name != db_name:
-                        entry["name"] = db_config.logic_name
                 else:
                     entry = {k: v for k, v in db_config.to_dict().items() if v}
                     # Remove internal fields
-                    for key in ("logic_name", "path_pattern", "extra", "default"):
+                    for key in ("path_pattern", "extra", "default"):
                         entry.pop(key, None)
```
[the remark this answers is not in the document]
Comment (liuyufei9527) [reply datus/cli/service_manager.py:292]: Skipping in this PR. `DbConfig.to_dict()` is `asdict()`, so all real connection fields (host/port/username/password/database/uri/schema/catalog) are top-level and survive serialization — `extra` only holds unknown/adapter-specific keys. Dropping `extra` on save is pre-existing behaviour (this PR only removed the now-deleted `logic_name` from that exclusion list), unrelated to the multi-database routing goal of this PR. If we want to round-trip adapter `extra` we should do it as a separate, focused change with its own tests.

File: datus/tools/db_tools/db_manager.py [hunk 2]
```diff
@@ -170,117 +171,100 @@ def _port_or_none(port_value: Optional[Union[str, int]]) -> Optional[int]:
...
+            match = next((f for f in files if f["name"] == target), None)
+            if match is None:
+                raise DatusException(
+                    ErrorCode.COMMON_VALIDATION_FAILED,
+                    message=f"Database '{target}' is not in datasource '{datasource}'. "
+                    f"Available: {', '.join(f['name'] for f in files)}.",
+                )
+            return match["name"], dataclasses.replace(cfg, uri=match["uri"], database=match["name"], path_pattern="")
+        if database and database != cfg.database:
+            return database, dataclasses.replace(cfg, database=database)
```
[the remark this answers is not in the document]
Comment (liuyufei9527) [reply datus/tools/db_tools/db_manager.py:255]: Acknowledged as a known limitation rather than fixing here. This PR's target is file-based (sqlite/duckdb glob) datasources, which go through the `path_pattern` branch where the URI is already rewritten per matched file. Same-datasource multi-database switching for *server* DBs configured purely via a URI is an edge case, and a generic URI-segment rewrite is dialect-specific and fragile. I'd prefer to handle server-side multi-DB URI rewriting in a dedicated follow-up with adapter-aware URI handling and tests, instead of a best-effort string patch here.

File: datus/tools/func_tool/semantic_tools.py [hunk 3]
```diff
@@ -275,7 +275,7 @@ def _extract_db_config(self, datasource: str) -> Optional[dict]:
         db_config = {
             k: str(v)
             for k, v in raw.items()
-            if v is not None and v != "" and k not in ("extra", "logic_name", "path_pattern", "catalog", "default")
+            if v is not None and v != "" and k not in ("extra", "path_pattern", "catalog", "default")
         }
```
[the remark this answers is not in the document]
Comment (liuyufei9527) [reply datus/tools/func_tool/semantic_tools.py:279]: No longer applicable — `logic_name` has been removed from the codebase entirely (it is not a field on `DbConfig` and never appears in `to_dict()`/`raw`), so the comprehension can't forward it. No change needed.

**Comment by liuyufei9527** [comment]
**Re: the out-of-diff scope-enforcement comments** (`database.py` ~983-1003, 1128-1149, 1090-1153 — `_check_sql_table_scope` using the primary connector's dialect / default database).

These are real but **pre-existing** and out of scope for this PR:
- The dialect used by `_check_sql_table_scope` only varies by *datasource*, and cross-datasource routing already existed before this PR; this PR only adds the *database* dimension (same datasource ⇒ same dialect), so the dialect concern predates it.
- Threading the routed `datasource`/`database` through the three scope call sites plus the coordinate-default resolution is a meaningful refactor of the scope subsystem.

I'd rather address scoped-tool routing fidelity in a dedicated follow-up (with adversarial scope-bypass tests) than bolt it onto this routing PR. Tracking it separately.

Summary of this round: fixed the DuckDB `schema_name` leak, `get_connections()` map contract, close()/cache-rebuild bug, `DatusException` pass-through, glob scope filtering, switch-database guard for single-file embedded DBs, redundant tool rebuild, plus test fixes (hermetic glob test, factory teardown restore, type hints, and multi-database dispatch coverage). Deferred/declined: `extra` serialization (pre-existing), server-URI rewrite (edge case), `logic_name` filter (field removed), and the scope items above.
